Fix iOS + Android WhatsApp Export Parsing

docflex opened the pull request:
### 🛠️ Fix iOS + Android WhatsApp Export Parsing

This PR addresses the parsing issues when reading WhatsApp chat exports from both **iOS** and **Android** devices. Specifically, it introduces normalization and regex updates to handle:

- ✅ Square brackets (`[]`) around iOS timestamps
- ✅ Narrow no-break space (`\u202f`) before AM/PM in iOS exports
- ✅ Left-to-right mark (`\u200E`) and other invisible characters
- ✅ Optional `~` between timestamp and sender (seen in Android exports)
- ✅ One-digit day/month values that previously failed regex matching

### 🔧 Changes Made

- Added normalization for invisible and platform-specific characters.
- Updated the regex to support both 1- and 2-digit day/month values.
- Made the sender separator (`~`) optional.
- Ensured the pattern works with or without seconds in timestamps.

This should make the `read_whatsapp_chat()` function much more robust and compatible across different export styles.

---

📎 **Linked Issue**: [[#6](https://github.com/ImadSaddik/Train_Your_Language_Model_Course/issues/6)]

docflex committed:
Parser Fix for Following:

- Square brackets ([]) around iOS timestamps
- Narrow no-break space (\u202f) before AM/PM in iOS
- Left-to-right mark (\u200E) invisible characters
- Optional "~" between timestamp and sender (Android)
- Single-digit day/month values not matching the regex

GitHub Issue: https://github.com/ImadSaddik/Train_Your_Language_Model_Course/issues/6

docflex committed:
Fixed Regex for Pattern Matching + Dummy Data To Check Both Types

- Android + iOS Type Exports for WhatsApp
- Pattern Matching Regex Fix

ImadSaddik commented on notebooks/1_DataCleaning.ipynb:
```diff
@@ -97,22 +96,32 @@
+    "    # 3. Remove Left-to-Right Mark (LRM) and Right-to-Left Mark (RTM) invisible characters\n",
     "    content = '\\n'.join(filtered_lines)\n",
+    "    content = content.replace('\\u202f', ' ')  # Replace narrow no-break space (iOS specific)\n",
+    "    content = content.replace('[', '').replace(']', '')  # Remove square brackets (iOS specific)\n",
```
Hi @docflex, you need to make sure that you replace only the square brackets that surround the timestamp. Currently, you are replacing any square bracket that might be in the text.

docflex replied:
Done with this. Please review in PR.

ImadSaddik commented on notebooks/1_DataCleaning.ipynb:
```diff
@@ -97,22 +96,32 @@
     "    for timestamp in df['timestamp']:\n",
     "        try:\n",
-    "            timestamp = pd.to_datetime(timestamp, format='mixed', dayfirst=True)\n",
+    "            timestamp = pd.to_datetime(timestamp, dayfirst=True, errors='coerce')\n",
```
Make sure to pull the latest changes because this line was changed. We use `format='mixed'` to let Pandas handle dates in every format possible.

docflex replied:
Done with this. Please review in PR.

ImadSaddik commented on notebooks/1_DataCleaning.ipynb:
```diff
@@ -97,22 +96,32 @@
     "    content = '\\n'.join(filtered_lines)\n",
+    "    content = content.replace('\\u202f', ' ')  # Replace narrow no-break space (iOS specific)\n",
+    "    content = content.replace('[', '').replace(']', '')  # Remove square brackets (iOS specific)\n",
+    "    content = content.replace('\\u200E', '').replace('\\u200F', '')  # Remove LRM and RTM characters\n",
```
Let's not remove these invisible formatting characters because they won't interfere with the regex pattern.

docflex replied:
These artifacts do end up in the final dataset if we don’t explicitly clean them at this stage. Since they provide no semantic value to the data we’re extracting, I believe they should be considered an essential part of the data cleaning process.

Additionally, they interfere with the regex logic used to isolate timestamps, leading to inaccurate parsing.

Here’s a screenshot illustrating the issue:

<img width="630" alt="Screenshot 2025-04-22 at 7 27 44 PM" src="https://github.com/user-attachments/assets/a544d9c5-66d8-4efc-9fac-d629f367e824" />

ImadSaddik replied:
I see, yes you are correct, we don't need those characters in the training set.

ImadSaddik commented on notebooks/1_DataCleaning.ipynb:
```diff
@@ -97,22 +96,32 @@
     "\n",
-    "    # Regular expression to match both WhatsApp message formats (with and without am/pm)\n",
-    "    pattern = r'(\\d{2}/\\d{2}/\\d{2,4}, \\d{1,2}:\\d{2}(?:\\s?[ap]m)?) - (.*?): (.*?)(?=\\n\\d{2}/\\d{2}/\\d{2,4}, \\d{1,2}:\\d{2}(?:\\s?[ap]m)? -|$)'\n",
+    "    # Normalize content:\n",
+    "    # 1. Remove any square brackets around timestamps (for iOS exports)\n",
+    "    # 2. Replace narrow no-break space (used in iOS exports) with a regular space\n",
+    "    # 3. Remove Left-to-Right Mark (LRM) and Right-to-Left Mark (RTM) invisible characters\n",
```
Update these comments accordingly.

docflex replied:
Done. Updated it in MD to reduce Code Clutter

ImadSaddik commented on notebooks/1_DataCleaning.ipynb:
```diff
@@ -97,22 +96,32 @@
     "            filtered_lines.append(line)\n",
     "\n",
-    "    # Regular expression to match both WhatsApp message formats (with and without am/pm)\n",
-    "    pattern = r'(\\d{2}/\\d{2}/\\d{2,4}, \\d{1,2}:\\d{2}(?:\\s?[ap]m)?) - (.*?): (.*?)(?=\\n\\d{2}/\\d{2}/\\d{2,4}, \\d{1,2}:\\d{2}(?:\\s?[ap]m)? -|$)'\n",
```
Update the regex to account for my other comments. Here is one that matches for the square brackets `[` and `]` at the beginning and end of the timestamp part.

```python
pattern = r'\[?(\d{2}/\d{2}/\d{2,4}, \d{1,2}:\d{2}(?:\s?[apAP][mM])?)\]?\s*[~-]\s*(.*?): (.*?)(?=\n\[?\d{2}/\d{2}/\d{2,4}, \d{1,2}:\d{2}(?:\s?[apAP][mM])?\]?\s*[~-]\s*|$)'
```

I have tried and it worked well.

docflex replied:
This does not seem to be working for iOS-type ones. I have added a few in the DummyData Against which you can run and verify. 

It doesn't work for iOS-type:

<img width="986" alt="Screenshot 2025-04-22 at 7 40 42 PM" src="https://github.com/user-attachments/assets/7885f026-6e86-4602-93a3-b55f82416ecc" />

ImadSaddik replied:
Interesting, I have tested on those messages, and I was able to extract the messages. I am not sure what happened in your case.

ImadSaddik commented on notebooks/1_DataCleaning.ipynb:
```diff
@@ -155,9 +164,20 @@
-   "outputs": [],
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "1345016"
+      ]
```
Don't forget to clean the output of all cells.

docflex replied:
Done.

ImadSaddik reviewed (changes requested):
The changes you made fixed the earlier issue. We can make it even better 😎, look at my comments.

docflex committed:
[RESOLVING COMMENTS]:

- Normalization Logic Streamlined
- Square Bracket Matching Fix
- Timestamp Pandas Fix

ImadSaddik commented:
@docflex thank you brother for the work you did on this MR ❤️. I merged it.

docflex commented:
Glad that I could be of service! 😉
